src/Agda2Hs/Compile/Function.hs

```diff
@@ -116,50 +115,50 @@ compileFun withSig def@Defn{..} =
     $ compileFun' withSig def
 
 -- inherit existing (instantiated) locals
-compileFun' withSig def@Defn{..} = do
+compileFun' withSig def@Defn{..} = inTopContext $ do
```

Comment:
It's mostly because of calls to `compileFun'` for compiling pattern-matching lambdas, which can appear in arbitrary contexts but should really be compiled in the top context. For `where` clauses we also call it from the context of the clause, which should (ideally) be the same as the context of the definition's module but we have to be careful not to add it twice. But I'll see if I can add the call to `inTopContext` to those places so we can avoid the blanket call here.

Reply:
Issue: https://github.com/agda/agda2hs/issues/306